chalice/deploy/packager.py

```diff
@@ -208,3 +205,258 @@ def _add_chalice_lib_if_needed(self, project_dir, zip_fileobj):
+        # type: () -> int
+        return hash(self.identifier)
+
+    def _calculate_identifier(self):
+        # type: () -> str
+        # Identiifer that can be fed into pip ie "name==version"
+        if self.dist_type == 'whl':
+            # From the wheel spec (PEP 427)
+            # {distribution}-{version}(-{build tag})?-{python tag}-{abi tag}-
+            # {platform tag}.whl
+            name, version = self.filename.split('-')[:2]
+        else:
+            name, version = self._parse_sdist_filename(self.filename)
+        return '%s==%s' % (name, version)
+
+    def _parse_sdist_filename(self, filename):
+        # type: (str) -> Tuple[str, str]
+        # Filenames are in the format {name}-{version}.tar.gz OR
+        # {name}-{version}.zip. These are the only two formats supported by
+        # PyPi for source distributions.
+        for ext in self.PYPI_SDIST_EXTS:
+            if filename.endswith(ext):
+                name, version = filename[:-len(ext)].split('-')
+                return name, version
+        raise InvalidSourceDistributionNameError(filename)
+
+
+class PipRunner(object):
+    """Wrapper around pip calls."""
+
+    def __init__(self, pip_module):
+        # type: (Any) -> None
+        self._pip = pip_module
+
+    @contextlib.contextmanager
+    def _consume_stdout_and_stderr(self):
+        # type: () -> Any
+        try:
+            out, err = sys.stdout, sys.stderr
+            temp_out, temp_err = StringIO(), StringIO()
+            sys.stdout, sys.stderr = temp_out, temp_err
+            yield temp_out, temp_err
+        finally:
+            sys.stdout, sys.stderr = out, err
+
+    def _execute(self, command, args):
+        # type: (str, List[str]) -> int
+        """Executes a pip command with the given arguments.
+
+        As an implementation detail this class assumes that pip was imported
+        and uses pip.main. This gives us the correct version of python that
+        the user is intending to deploy for free, but relies on the pip main
+        function not changing, which is very unlikely.
+        """
+        final_command = [command]
+        final_command.extend(args)
+        # Call pip and hide stdout/stderr
+        with self._consume_stdout_and_stderr():
+            rc = self._pip.main(final_command)
+        return rc
+
+    def build_wheel(self, wheel, directory):
+        # type: (str, str) -> int
+        """This command builds an sdist into a wheel file."""
+        arguments = ['--no-deps', '--wheel-dir', directory, wheel]
+        return self._execute('wheel', arguments)
+
+    def download_all_dependencies(self, requirements_file, directory):
+        # type: (str, str) -> int
+        """This command downloads all dependencies as sdists."""
+        # None of these should fail so we can request non-binary versions of
+        # all the packages in one call to `pip download`
+        arguments = ['--no-binary=:all:', '-r', requirements_file, '--dest',
```

Comment:
Hmmm interesting. I guess technically pip is an implementation detail of the requirements builder. Though I think that would be useful to allow people who are getting requirements that are not on pypi.